lib/Dialect/AIG/Transforms/BalanceVariadic.cpp

```diff
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass lowers variadic AndInverter operations to balanced binary
+// AndInverter operations.
+//
+//===----------------------------------------------------------------------===//
+#include "llvm/ADT/PriorityQueue.h"
+
+#include "circt/Dialect/AIG/AIGOps.h"
+#include "circt/Dialect/AIG/AIGPasses.h"
+#include "circt/Dialect/AIG/Analysis/OpDepthAnalysis.h"
+
+#define DEBUG_TYPE "aig-balance-variadic"
+
+namespace circt {
+namespace aig {
+#define GEN_PASS_DEF_BALANCEVARIADIC
+#include "circt/Dialect/AIG/AIGPasses.h.inc"
+} // namespace aig
+} // namespace circt
+
+using namespace circt;
+using namespace aig;
+
+namespace {
+/// For wrapping Value and complement information into one object
+struct Signal {
+  Value value;
+  bool complement;
+
+  Signal() = default;
+  Signal(Value v, bool complement) : value(v), complement(complement) {}
+
+  bool isComplement() const { return complement; }
+  Value getValue() const { return value; }
+};
+
+struct BalanceVariadicDriver {
+  BalanceVariadicDriver(mlir::IRRewriter &rewriter,
+                        aig::analysis::OpDepthAnalysis *opDepthAnalysis)
+      : rewriter(rewriter), opDepthAnalysis(opDepthAnalysis) {}
+
+  struct PairSorter {
+    bool operator()(const std::pair<size_t, Signal> &lhs,
+                    const std::pair<size_t, Signal> &rhs) const {
+      // First compare by level (higher level = lower priority)
+      if (lhs.first != rhs.first)
+        return lhs.first > rhs.first;
+
+      // If levels are equal, compare by argnumber or result number for
+      // deterministic ordering
+      auto *lop = lhs.second.getValue().getDefiningOp();
+      auto *rop = rhs.second.getValue().getDefiningOp();
+      if (lop && rop) {
+        return !lop->isBeforeInBlock(rop);
+      }
+
+      if (!lop && rop)
+        return false;
+
+      if (lop && !rop)
+        return true;
+
+      BlockArgument larg = cast<BlockArgument>(lhs.second.getValue());
+      BlockArgument rarg = cast<BlockArgument>(rhs.second.getValue());
+      return larg.getArgNumber() > rarg.getArgNumber();
+    }
+  };
+
+  using NodeLevelHeap =
+      llvm::PriorityQueue<std::pair<size_t, Signal>,
+                          std::vector<std::pair<size_t, Signal>>, PairSorter>;
+
+  void balanceVariadicAndInverterOp(AndInverterOp op) {
+    rewriter.setInsertionPoint(op);
+
+    NodeLevelHeap sortByLevel;
+    for (auto [fanin, inverted] :
+         llvm::zip(op.getOperands(), op.getInverted())) {
+      auto faninOp = fanin.getDefiningOp<AndInverterOp>();
+      size_t level = faninOp ? opDepthAnalysis->updateLevel(faninOp, true) : 0;
+      sortByLevel.push({level, Signal(fanin, inverted)});
+    }
+
+    // extract the top two elements with minimum level
+    // and replace them with a new AndInverterOp
+    while (sortByLevel.size() > 2) {
```

Comment:
nit: Can you change the condition to `sortByLevel.size() > 1` and remove the default case in the following switch statement? I think it's a bit redundant.